// @cs-au-dk/jelly/lib/analysis/accesspaths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownAccessPath = exports.IgnoredAccessPath = exports.ComponentAccessPath = exports.CallResultAccessPath = exports.PropertyAccessPath = exports.ModuleAccessPath = exports.AccessPath = void 0;
const infos_1 = require("./infos");
class AccessPath {
    str;
    constructor(str) {
        this.str = str;
    }
    toString() {
        return this.str;
    }
}
exports.AccessPath = AccessPath;
class ModuleAccessPath extends AccessPath {
    moduleInfo;
    requireName;
    constructor(moduleInfo, requireName) {
        const t = !"./#".includes(requireName[0]) && requireName !== moduleInfo.getOfficialName() ? requireName : undefined;
        super(`<${moduleInfo.getOfficialName()}${moduleInfo instanceof infos_1.ModuleInfo ? `@${moduleInfo.packageInfo.version}` : ''}${t ? `(${t})` : ""}>`);
        this.moduleInfo = moduleInfo;
        this.requireName = t;
    }
}
exports.ModuleAccessPath = ModuleAccessPath;
class PropertyAccessPath extends AccessPath {
    base;
    prop;
    constructor(base, prop) {
        super(`${base}.${prop}`);
        this.base = base;
        this.prop = prop;
    }
}
exports.PropertyAccessPath = PropertyAccessPath;
class CallResultAccessPath extends AccessPath {
    caller;
    constructor(caller) {
        super(`${caller}()`);
        this.caller = caller;
    }
}
exports.CallResultAccessPath = CallResultAccessPath;
class ComponentAccessPath extends AccessPath {
    component;
    constructor(component) {
        super(`${component}<>`);
        this.component = component;
    }
}
exports.ComponentAccessPath = ComponentAccessPath;
class IgnoredAccessPath extends AccessPath {
    static instance = new IgnoredAccessPath();
    constructor() {
        super("Ignored");
    }
}
exports.IgnoredAccessPath = IgnoredAccessPath;
class UnknownAccessPath extends AccessPath {
    static instance = new UnknownAccessPath();
    constructor() {
        super("Unknown");
    }
}
exports.UnknownAccessPath = UnknownAccessPath;
