// @cs-au-dk/jelly/lib/analysis/listeners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenListener = void 0;
var TokenListener;
(function (TokenListener) {
    TokenListener[TokenListener["ANCESTORS"] = 0] = "ANCESTORS";
    TokenListener[TokenListener["CALL_FUNCTION_BASE"] = 1] = "CALL_FUNCTION_BASE";
    TokenListener[TokenListener["CALL_FUNCTION_ANCESTORS"] = 2] = "CALL_FUNCTION_ANCESTORS";
    TokenListener[TokenListener["CALL_FUNCTION_CALLEE"] = 3] = "CALL_FUNCTION_CALLEE";
    TokenListener[TokenListener["CALL_FUNCTION_EXTERNAL"] = 4] = "CALL_FUNCTION_EXTERNAL";
    TokenListener[TokenListener["READ_PROPERTY_BASE"] = 5] = "READ_PROPERTY_BASE";
    TokenListener[TokenListener["READ_PROPERTY_GETTER"] = 6] = "READ_PROPERTY_GETTER";
    TokenListener[TokenListener["READ_PROPERTY_GETTER2"] = 7] = "READ_PROPERTY_GETTER2";
    TokenListener[TokenListener["READ_PROPERTY_GETTER_THIS"] = 8] = "READ_PROPERTY_GETTER_THIS";
    TokenListener[TokenListener["READ_PROPERTY_GETTER_THIS2"] = 9] = "READ_PROPERTY_GETTER_THIS2";
    TokenListener[TokenListener["READ_PROPERTY_BASE_DYNAMIC"] = 10] = "READ_PROPERTY_BASE_DYNAMIC";
    TokenListener[TokenListener["PACKAGE_NEIGHBORS"] = 11] = "PACKAGE_NEIGHBORS";
    TokenListener[TokenListener["READ_ANCESTORS"] = 12] = "READ_ANCESTORS";
    TokenListener[TokenListener["CLASS_FIELD"] = 13] = "CLASS_FIELD";
    TokenListener[TokenListener["EXTENDS"] = 14] = "EXTENDS";
    TokenListener[TokenListener["ASSIGN_MEMBER_BASE"] = 15] = "ASSIGN_MEMBER_BASE";
    TokenListener[TokenListener["ASSIGN_SETTER"] = 16] = "ASSIGN_SETTER";
    TokenListener[TokenListener["ASSIGN_SETTER_THIS"] = 17] = "ASSIGN_SETTER_THIS";
    TokenListener[TokenListener["ASSIGN_DYNAMIC_BASE"] = 18] = "ASSIGN_DYNAMIC_BASE";
    TokenListener[TokenListener["ASSIGN_OBJECT_PATTERN_REST"] = 19] = "ASSIGN_OBJECT_PATTERN_REST";
    TokenListener[TokenListener["ASSIGN_OBJECT_PATTERN_REST_PROPERTIES"] = 20] = "ASSIGN_OBJECT_PATTERN_REST_PROPERTIES";
    TokenListener[TokenListener["ASSIGN_ARRAY_PATTERN_REST"] = 21] = "ASSIGN_ARRAY_PATTERN_REST";
    TokenListener[TokenListener["ASSIGN_ARRAY_PATTERN_REST_ARRAY"] = 22] = "ASSIGN_ARRAY_PATTERN_REST_ARRAY";
    TokenListener[TokenListener["ASSIGN_REQUIRE_EXTENSIONS"] = 23] = "ASSIGN_REQUIRE_EXTENSIONS";
    TokenListener[TokenListener["ASSIGN_ANCESTORS"] = 24] = "ASSIGN_ANCESTORS";
    TokenListener[TokenListener["READ_ITERATOR_VALUE"] = 25] = "READ_ITERATOR_VALUE";
    TokenListener[TokenListener["IMPORT_BASE"] = 26] = "IMPORT_BASE";
    TokenListener[TokenListener["EXPORT_BASE"] = 27] = "EXPORT_BASE";
    TokenListener[TokenListener["OBJECT_SPREAD"] = 28] = "OBJECT_SPREAD";
    TokenListener[TokenListener["NATIVE_INVOKE_CALLBACK"] = 29] = "NATIVE_INVOKE_CALLBACK";
    TokenListener[TokenListener["NATIVE_INVOKE_CALLBACK2"] = 30] = "NATIVE_INVOKE_CALLBACK2";
    TokenListener[TokenListener["NATIVE_INVOKE_CALL_APPLY2"] = 31] = "NATIVE_INVOKE_CALL_APPLY2";
    TokenListener[TokenListener["NATIVE_INVOKE_CALL_APPLY3"] = 32] = "NATIVE_INVOKE_CALL_APPLY3";
    TokenListener[TokenListener["CALL_PROMISE_EXECUTOR"] = 33] = "CALL_PROMISE_EXECUTOR";
    TokenListener[TokenListener["CALL_PROMISE_RESOLVE"] = 34] = "CALL_PROMISE_RESOLVE";
    TokenListener[TokenListener["CALL_PROMISE_ONFULFILLED"] = 35] = "CALL_PROMISE_ONFULFILLED";
    TokenListener[TokenListener["CALL_PROMISE_ONREJECTED"] = 36] = "CALL_PROMISE_ONREJECTED";
    TokenListener[TokenListener["CALL_PROMISE_ONFINALLY"] = 37] = "CALL_PROMISE_ONFINALLY";
    TokenListener[TokenListener["MAKE_PROMISE_RESOLVE"] = 38] = "MAKE_PROMISE_RESOLVE";
    TokenListener[TokenListener["MAKE_PROMISE_REJECT"] = 39] = "MAKE_PROMISE_REJECT";
    TokenListener[TokenListener["MAKE_PROMISE_ALL"] = 40] = "MAKE_PROMISE_ALL";
    TokenListener[TokenListener["MAKE_PROMISE_ALLSETTLED"] = 41] = "MAKE_PROMISE_ALLSETTLED";
    TokenListener[TokenListener["MAKE_PROMISE_ANY"] = 42] = "MAKE_PROMISE_ANY";
    TokenListener[TokenListener["MAKE_PROMISE_RACE"] = 43] = "MAKE_PROMISE_RACE";
    TokenListener[TokenListener["AWAIT"] = 44] = "AWAIT";
    TokenListener[TokenListener["JSX_ELEMENT"] = 45] = "JSX_ELEMENT";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES"] = 46] = "NATIVE_ASSIGN_PROPERTIES";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES2"] = 47] = "NATIVE_ASSIGN_PROPERTIES2";
    TokenListener[TokenListener["NATIVE_ASSIGN_PROPERTIES3"] = 48] = "NATIVE_ASSIGN_PROPERTIES3";
    TokenListener[TokenListener["NATIVE_OBJECT_CREATE"] = 49] = "NATIVE_OBJECT_CREATE";
    TokenListener[TokenListener["NATIVE_OBJECT_DEFINE_PROPERTY"] = 50] = "NATIVE_OBJECT_DEFINE_PROPERTY";
    TokenListener[TokenListener["NATIVE_OBJECT_DEFINE_PROPERTIES"] = 51] = "NATIVE_OBJECT_DEFINE_PROPERTIES";
    TokenListener[TokenListener["NATIVE_8"] = 52] = "NATIVE_8";
    TokenListener[TokenListener["NATIVE_11"] = 53] = "NATIVE_11";
    TokenListener[TokenListener["NATIVE_12"] = 54] = "NATIVE_12";
    TokenListener[TokenListener["NATIVE_14"] = 55] = "NATIVE_14";
})(TokenListener || (exports.TokenListener = TokenListener = {}));
